### ISSUE TITLE
buf fails to parse a proto file with options that use a group (which protoc supports)

### ISSUE THREAD

**Issue by eyalpost** [bufbuild/buf#620]
We have some proto files using the group construct which pass when using `protoc` but fail when using `buf protoc`.
To reproduce:

`options.proto`:
```proto
syntax = "proto2";

import "google/protobuf/descriptor.proto";

message Composite {
  optional string str = 1001;
  optional group CompositeGroup = 1006 {
    required int32 grouped_int32 = 1;
    optional string grouped_str = 2;
  }
}

extend google.protobuf.MessageOptions {
  optional Composite compo = 30000;
}
```

`service.proto`:
```proto
syntax = "proto3";

import "options.proto";

message Message {
  option (compo) = {
    str: "a string"
    CompositeGroup: {
      grouped_int32: 1
      grouped_str: "hello"
    }
  };
}
```

running `protoc` vs `buf protoc` provides the following:
(Note that the referenced `.. ../include` folder contains the `google/protobuf/descriptor.proto` stuff)
```bash
> protoc -I . service.proto -I ../../../../../include  -o /dev/null
👍 
> buf protoc -I . service.proto -I ../../../../../include  -o /dev/null
service.proto:6:20:message Message: option (compo): field CompositeGroup not found
```
Interestingly enough, If I change the `service.proto` file to use `compositegroup` instead of `CompositeGroup`,  i.e.
```proto
message Message {
  option (compo) = {
    str: "a string"
    compositegroup: {
      grouped_int32: 1
      grouped_str: "hello"
    }
  };
}
```

buf completes successfully but protoc fails:
```bash
> buf protoc -I . service.proto -I ../../../../../include  -o /dev/null
👍 
> protoc -I . service.proto -I ../../../../../include  -o /dev/null
service.proto:6:20: Error while parsing option value for "compo": Message type "Composite" has no field named "compositegroup".
```

Seems like buf expects the "internal" name of the group (which is lowered case `compositegroup`) but apparently that's not the supported proto text format

**Comment by jhump** [comment]
Oof, what a crazy find!

Both buf and protoc actually generate the field (in the descriptor) with the lower-case name. And that's why buf is expecting the lower-case name in the text format.

This is kind of a shame that protoc works this way with aggregates. Because protoc actually works just like buf if you _destructure_ the message into multiple options:

```proto
syntax = "proto2";
  
import "google/protobuf/descriptor.proto";

message Foo {
  optional group Bar = 1 {
    optional string name = 1;
    optional uint64 id = 2;
    repeated string labels = 3;
  }
}

extend google.protobuf.MessageOptions {
  optional Foo foo = 10001;
  optional group Baz = 10002 {
    optional string x = 1;
    optional string y = 2;
  }
}

message Frobnitz {
  // protoc chokes if you use "Bar" here, which is how the group name is defined
  option (foo).bar.name = "abc";
  // ditto that protoc chokes if you use "Baz" here
  option (baz).x = "xyz";

  repeated int32 vals = 1;
}

message Gyzmeaux {
  option (foo) = {
    // protoc requires "Bar" here and chokes on "bar"
    Bar<
      name: "abc"
      id: 1234
    >
  };
}
```

This almost feels like a protoc bug because if I use `Bar` (capitalized) in the first option, I get this error:
```
test.proto:23:10: Option "(foo).Bar" unknown. Ensure that your proto definition file imports the proto which defines the option.
```
But if I use `bar` (lower-case) in the last option, I get this error:
```
test.proto:31:18: Error while parsing option value for "foo": Message type "Foo" has no field named "bar".
```
These errors seem incompatible, like indicative of a protoc bug... However, since groups are a very old feature that aren't even supported by proto3, I have no doubt that the protobuf team would not entertain the idea of changing protoc here.

Anyhow, I know where to fix this to make buf match protoc.

**Comment by jhump** [comment]
Eek, more evidence that this is really a long-standing bug in protoc (that I guess we'll just have to reproduce because protoc is the standard, despite the warts!): protoc does _not_ work this way for _extensions_ that are groups:

```proto
syntax = "proto2";
  
import "google/protobuf/descriptor.proto";

message Foo {
  extensions 1 to 10;
}

extend Foo {
  optional group Bar = 10 {
    optional string a = 1;
    optional string b = 2;
  }
}

extend google.protobuf.MessageOptions {
  optional Foo foo = 10000;
}

message Baz {
  option (foo) = {
    // even though a regular field would require the group name
    // (not all lower-cased name), for extensions, protoc requires
    // the lower-cased name and chokes if we used "Bar" :(
    [Bar]<
      a: "a1"
      b: "b2"
    >
  };
}
```

If we try to use the group name (capitalized), we get an error:
```
test.proto:21:18: Error while parsing option value for "foo": Extension "Bar" is not defined or is not an extension of "Foo".
```

**Comment by eyalpost** [comment]
yes, I kinda assumed this is a protoc bug in the first place since these CamelCase field names seemed out of place 🙂 
Thanks for your efforts!

**Comment by bufdev** [comment]
This is fixed on `main` and will go out in the next release. Thanks for the bug report!

**Comment by bufdev** [comment]
This is fixed on v1.0.0-rc3.
